gtdynamics/dynamics/ChainDynamicsGraph.cpp

```diff
@@ -0,0 +1,223 @@
+
+    // Create expression for wrench constraint on trunk
+    Vector6_ trunk_wrench_constraint = gravity_wrench;
+
+    // Set tolerances
+
+    // Get tolerance
+    Vector6 wrench_tolerance = opt().f_cost_model->sigmas();
+    Vector3 contact_tolerance = opt().cm_cost_model->sigmas();
+
+     std::vector<Key> wrench_keys;
+
+    for (int i = 0 ; i < 4 ; ++i) {
+    
+    bool foot_in_contact = false;
+    // Get key for wrench at hip joint with id 0
+    const Key wrench_key_3i_T = WrenchKey(0, 3*i, t);
+
+    // create expression for the wrench and initialize to zero
+    Vector6_ wrench_3i_T(Vector6::Zero());
+
+    // Set wrench expression on trunk by leg, according to contact 
+    for (auto&& cp : *contact_points) {
+      if (cp.link->id() == 3*(i+1)) {
+        wrench_3i_T = Vector6_(wrench_key_3i_T);
+        foot_in_contact = true;
+      }
+    }
+
+    // add wrench to trunk constraint
+    wrench_keys.push_back(wrench_key_3i_T);
+    
+
+    // Get expression for end effector wrench using chain
+    Vector6_ wrench_end_effector =
+        composed_chains_[i].AdjointWrenchConstraint3(chain_joints_[i], wrench_key_3i_T, t);
+
+    // Create contact constraint
+    Point3 contact_in_com(0.0, 0.0, -0.07);
+    Vector3_ contact_constraint = ContactDynamicsMomentConstraint(
+        wrench_end_effector, gtsam::Pose3(gtsam::Rot3(), (-1) * contact_in_com));
+    auto contact_expression = VectorExpressionEquality<3>(contact_constraint,
+                                                           contact_tolerance);
+    if (foot_in_contact)
+    graph.add(contact_expression.createFactor(1));
+    else {
+      Vector6 wrench_zero = gtsam::Z_6x1;
+      graph.addPrior(wrench_key_3i_T, wrench_zero, opt().f_cost_model);
+    }
+  }
+
+  // Add wrench factor for trunk link
+    graph.add(
+          WrenchFactor(opt().f_cost_model, robot.link("trunk"), wrench_keys, t, gravity()));
+
+  return graph;
+}
+
+gtsam::NonlinearFactorGraph ChainDynamicsGraph::qFactors(
+    const Robot &robot, const int t,
+    const std::optional<PointOnLinks> &contact_points) const {
+  NonlinearFactorGraph graph;
+
+  // Get Pose key for base link
+  const Key base_key = PoseKey(0, t);
+
+  // Get tolerance
+  Vector6 tolerance = opt().p_cost_model->sigmas();
+
+  for (int i = 0; i < 4; ++i) {
+    // Get end effector key
+    const Key end_effector_key = PoseKey(3*(i+1), t);
+
+    // Get expression for end effector pose
+    gtsam::Vector6_ chain_expression = composed_chains_[i].Poe3Factor(
+        chain_joints_[i], base_key, end_effector_key, t);
+
+    auto chain_constraint= VectorExpressionEquality<6>(chain_expression,
+                                                           tolerance);
+
+    graph.add(chain_constraint.createFactor(1.0));
+  }
+
+  gtsam::Vector3 gravity_used;
+  auto graph_gravity = gravity();
+  if (graph_gravity)
+    gravity_used = *graph_gravity;
+  else
+    gravity_used = gtsam::Vector3(0, 0, -9.8);
```

Comment:
This is so awkward. Why not just have `gravity_` set to `Vector3(0, 0, -9.8)` as the default? If someone passes in an alternate value, then that will get assigned directly, no need to create pointless copies of the gravity vector.

I think this is a problem in multiple places outside the scope of your PR. Please leave a TODO to clean this up and I can fix it later.